### PR TITLE
bzimage: add an extract function, tests work

### PR DESCRIPTION
**Pull request by rminnich**
Extract returns the kernel part of the bzimage, uncompressed.

Extract now works, tests now work, and the copy subcommand now works.
bzimage can now pull a bzimage apart, change nothing, and put it back
together and it works. This may not sound like much but ... it is.

Signed-off-by: Ronald G. Minnich <rminnich@gmail.com>

### CODE REVIEW

**Comment by rminnich** [comment]
I'm fairly sure the way I'm doing this is safe. There are lots of fixed requirements now on bzimage structure so I don't expect lots of future trouble.

And, amazingly, the copy command works.
bzimage copy /tmp/bzImage x
qemu .... -kernel x

boots just fine. There is one byte of difference which I'm trying to work out now.

**Comment by rjoleary** [comment]
Did you find out why that one byte was different?

**Comment by rminnich** [comment]
not yet. It may no longer be ... this last iteration fixed a LOT of problems.

**Comment by rminnich** [comment]
there's 3 bytes in a four byte sequence after the YZ magic that differ
bzimage
003b8900  9b e3 51 40 03 00 00 00  00 01 59 5a f8 85 21 01  |..Q@......YZ..!.|
                                                                                       ^^^^^^^^

copy
003b8900  9b e3 51 40 03 00 00 00  00 01 59 5a a8 f0 6e 01  |..Q@......YZ..n.|
                                                                                      ^^^^^^^^^

I'll work on it on the flight.

File: cmds/bzimage/bzimage.go [hunk 1]
```diff
@@ -46,20 +55,82 @@ func main() {
...
+		var br2 = &bzimage.BzImage{}
+		if err = br2.UnmarshalBinary(b2); err != nil {
+			log.Fatal(err)
+		}
+		fmt.Printf("%s", br.Header.Diff(&br2.Header))
 	case "dump":
 		fmt.Printf("%s\n", strings.Join(br.Header.Show(), "\n"))
+	case "extract":
+		bzimage.Debug = log.Printf
+		//fmt.Printf("%s\n", strings.Join(br.Header.Show(), "\n"))
```
Comment (rjoleary) [inline cmds/bzimage/bzimage.go]: Remove comment line.

File: pkg/bzimage/bzimage.go [hunk 2]
```diff
@@ -172,7 +334,119 @@ func (h *LinuxHeader) Diff(i *LinuxHeader) string {
...
+	}
+	return s
+}
+
 // String stringifies a LinuxHeader into comma-separated parts
 func (h *LinuxHeader) String() string {
 	return strings.Join(h.Show(), ",")
 }
+
+// InitRAMFS returns a []byte from KernelCode which can be used save or replace
```
Comment (rjoleary) [inline pkg/bzimage/bzimage.go]: used *to* save or replace

File: pkg/bzimage/bzimage.go [hunk 3]
```diff
@@ -7,27 +7,53 @@
...
 // For now, it hardwires the KernelBase to 0x100000.
+// bzImages were created by a process of evilution, and they are wondrous to behold.
+// bzImages are almost impossible to modify. They form a sandwich with
+// the compressed kernel code in the middle. It's actually a BLT:
+// MBR and bootparams first 512 bytes
+//    the MBR includes 0xc0 bytes of boot code which is vestigial.
+// Then there is "preamble" code which is the kernel decompressor; then the
+// xz compressed kernel; then a library of sorts after the kernel which is called
+// by the early uncompressed kernel code. This is all linked together and forms
+// an essentially indivisible whole -- which we wish to divisible.
```
[the remark this answers is not in the document]
Comment (rminnich) [reply pkg/bzimage/bzimage.go]: it's a joke

File: pkg/cpio/types.go [hunk 4]
```diff
@@ -23,6 +23,11 @@ type Record struct {
 
 	// Info is metadata describing the CPIO record.
 	Info
+
+	// metadata about this item's place in the file
+	RecPos  int64  // Where in the file this record is
+	RecLen  uint64 // How big the record is.
+	FilePos int64  // Where in the file the file is.
```
[the remark this answers is not in the document]
Comment (rminnich) [reply pkg/cpio/types.go]: done

File: pkg/bzimage/bzimage.go (hunk 2 above)
Comment (rminnich) [reply pkg/bzimage/bzimage.go]: done

File: pkg/cpio/newc.go [hunk 5]
```diff
@@ -239,11 +240,16 @@ func (r *reader) ReadRecord() (Record, error) {
 	info := hdr.Info()
 	info.Name = string(nameBuf[:hdr.NameLength-1])
 
+	RecLen := uint64(r.pos - RecPos)
+	FilePos := r.pos
```
[the remark this answers is not in the document]
Comment (rminnich) [reply pkg/cpio/newc.go]: done

File: pkg/cpio/newc.go [hunk 6]
```diff
@@ -239,11 +240,16 @@ func (r *reader) ReadRecord() (Record, error) {
 	info := hdr.Info()
 	info.Name = string(nameBuf[:hdr.NameLength-1])
 
+	RecLen := uint64(r.pos - RecPos)
```
[the remark this answers is not in the document]
Comment (rminnich) [reply pkg/cpio/newc.go]: done

File: pkg/cpio/newc.go [hunk 7]
```diff
@@ -206,6 +206,7 @@ func (r *reader) readAligned(p []byte) error {
 // ReadRecord implements RecordReader for the newc cpio format.
 func (r *reader) ReadRecord() (Record, error) {
 	hdr := header{}
+	RecPos := r.pos
```
[the remark this answers is not in the document]
Comment (rminnich) [reply pkg/cpio/newc.go]: done

File: pkg/bzimage/bzimage.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rminnich) [reply pkg/bzimage/bzimage.go]: I'll be so sad if I have to take it out.

**Comment by paulmenzel** [comment]
Have you looked at `[scripts/extract-vmlinux](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/scripts/extract-vmlinux)` from the Linux kernel?

File: cmds/bzimage/bzimage.go (hunk 1 above)
Comment (rminnich) [reply cmds/bzimage/bzimage.go]: done

**Comment by rminnich** [comment]
I looked at that script. I did not find it all that useful.

**Comment by rminnich** [comment]
in general linux kernel scripts all need to die.

**Pull request merged**